src/org/github/jamm/MemoryLayoutSpecification.java

```diff
@@ -1,144 +1,126 @@
-        final boolean hasDot = strSpecVersion.indexOf('.') != -1;
-        if (hasDot) {
-            if ("1".equals(strSpecVersion.substring(0, strSpecVersion.indexOf('.')))) {
-                // Java 1.6, 1.7, 1.8
-                final String strVmVersion = System.getProperty("java.vm.version");
-                if (strVmVersion.startsWith("openj9"))
-                {
-                    modernJvm = true;
-                }
-                else
-                {
-                    final int vmVersion = Integer.parseInt(strVmVersion.substring(0, strVmVersion.indexOf('.')));
-                    modernJvm = vmVersion >= 17;
-                }
-            }
-        }
-
-        final int alignment = getAlignment();
-        if (modernJvm) {
-
-            long maxMemory = 0;
-            for (MemoryPoolMXBean mp : ManagementFactory.getMemoryPoolMXBeans()) {
-                maxMemory += mp.getUsage().getMax();
-            }
+        if (VM.useCompressedOops()) {
 
-            if (maxMemory < 30L * 1024 * 1024 * 1024) {
-                // HotSpot 17.0 and above use compressed OOPs below 30GB of RAM
-                // total for all memory pools (yes, including code cache).
-                return new MemoryLayoutSpecification() {
-
-                    public int getArrayHeaderSize() {
-                        return 16;
-                    }
+            return new MemoryLayoutSpecification() {
 
-                    public int getObjectHeaderSize() {
-                        return 12;
-                    }
+                @Override
+                public int getArrayHeaderSize() {
+                    return 16; // object header (12 bytes) + array length (4 bytes)
+                }
 
-                    public int getObjectPadding() {
-                        return alignment;
-                    }
+                @Override
+                public int getObjectHeaderSize() {
+                    return 12; // mark word (8 bytes) + class word (4 bytes)
+                }
 
-                    public int getReferenceSize() {
-                        return 4;
-                    }
+                @Override
+                public int getObjectAlignment() {
+                    return objectAlignment;
+                }
 
-                    public int getSuperclassFieldPadding() {
-                        return 4;
-                    }
-                };
-            }
+                @Override
+                public int getReferenceSize() {
+                    return 4;
+                }
+            };
         }
 
-        /* Worst case we over count. */
-
         // In other cases, it's a 64-bit uncompressed OOPs object model
         return new MemoryLayoutSpecification() {
 
+            @Override
             public int getArrayHeaderSize() {
-                return 24;
+                return 20; // object header (16 bytes) + array length (4 bytes)
```

Comment:
In my understanding we only intentionally overcounted for fields. Due to field alignement it can happen that the JVM use less space than what our simple logic expect https://shipilev.net/jvm/objects-inside-out/#_observation_hiding_fields_in_field_alignment_gaps.

Reply:
With good test coverage we should be good

Reply:
Fix that one back in a proper way

Reply:
Yes, I would like to point out you not only fixed but also documented those things so next person does not have to rediscover them. Thank you for doing it :-)